[docs] Add 'Why Toolpad?' (#2632)
mui/mui-toolpad: docs/data/toolpad/getting-started/why-toolpad.md

# Why Toolpad?

> <p class="description">Explains our motivation for building Toolpad, where it differs from other tools and how it could be useful to you.</p>

Reviewer: Iterating on the suggestion above to: 1) center the needs of the reader (they probably don't really care about "our motivation") and 2) specify what kind of tools we're talking about

```suggestion
<p class="description">Learn how Toolpad differs from the other internal tool builders, and how it can be useful to you.</p>
```

## Introduction

Toolpad is primarily designed as an application framework for internal tool development. Such applications are often built by teams with access to fewer frontend development resources. These teams face challenges such as:

- Having to keep up with the fast evolving landscape of libraries, frameworks and best practices.

> - Lots of glue code is required to get backend data to the browser and hooked up to data displaying components.

Reviewer: Rephrasing for consistent verb tense across list items

```suggestion
- Dealing with all of the glue code required to get backend data to the browser and hooked up to components for display.
```

- As backends evolve, the internal applications need to evolve with them with minimal amount of maintenance required.

> - Lesser focus on quality assurance usually leads to higher chances of failure.

Reviewer: these don't feel like "challenges" in the same vein as the last two list items—feels like they're just restating the same core thesis here, which is that teams often don't have the resources to maintain the front end. Maybe this list is not necessary at all?

## TL;DR

- Integrates well with your existing code. You can reuse your database models, client libraries, secrets, bespoke components
- Integrates well with your development lifecycle. Your project lives on your own file system. You can use your own IDE, version control system, deployment target, CLI tools
- Comes with a drag-and-drop builder to simplify UI building
- Sets healthy UI constraints that prevent common pitfalls and anti-patterns
- Lets you maintain a single source of truth for all business logic
- Less boilerplate so you can focus on the important parts of the app

## Market overview

After reviewing multiple approaches and tools, we noticed that there are a few recurring core problem areas that needed to be addressed:

### The challenge of codebase fragmentation and duplication

> Frequently, developers invest valuable time in hunting down particular versions of business logic and then maintaining them at multiple places. When working with external tools, code duplication can often arise as consequences of design decisions. Not having a unified reference point within a codebase compromises its scalability, ease of maintenance and debugging.

Reviewer: ```suggestion
Frequently, developers invest valuable time in hunting down particular versions of business logic and then maintaining them in multiple places. When working with external tools, code duplication can often arise a consequence of design decisions. Not having a unified reference point within a codebase compromises its scalability as well as ease of maintenance and debugging.
```

### Loss of control with external tools

When working with external tools, developers often forfeit their autonomy and are constrained by the choices offered by the tool. Developers typically derive a sense of ownership and responsibility when they oversee the entire process, encompassing integrations, deployment, and version control. However, when a black-box element intervenes, control often shifts away.

### Trade-off between speed and customization

Typically, depending on factors such as project size, time constraints, available resources, and scope, developers are compelled to strike a balance between speed and customization. Every business wants features to be delivered at speed, but being able to customize later keeps the project open to come what may. While both aspects are significant, we believe that an architecturally well-designed tool empowers you to attain both objectives simultaneously.

> ### Trust and Safety

Reviewer: ```suggestion
### Trust and safety
```

> Internal tools have access to highly sensitive data, data that every organization solemnly pledges to safeguard. Despite the stringent security measures in place, data breaches remain a prevalent occurrence. Organizations rigorously engage in due diligence and compliance assessments before placing trust in a vendor. The security of the underlying data remains an exceptionally vital aspect that cannot be underestimated.

Reviewer: Rephrasing the last part a lil bit to try to make the point more clear.

```suggestion
Internal tools have access to highly sensitive data—data that every organization solemnly pledges to safeguard. Despite stringent security measures, data breaches remain a prevalent occurrence. Organizations rigorously engage in due diligence and compliance assessments before placing trust in a vendor. The security of the underlying data is critical, and the seriousness of this matter can't be overstated.
```

> ## Toolpad

Reviewer: ```suggestion
## How Toolpad fits in
```

Reviewer: or

```suggestion
## Where Toolpad comes in
```

> We think addressing the above mentioned problems can lead to a better developer experience. Toolpad solves them in the following ways:

Reviewer: No need to speculate about what you "think"—say it like you mean it! It's true if we say that it is. 😁

```suggestion
Addressing all the aforementioned problems leads to a better developer experience. Toolpad solves them in the following ways:
```

### Front-end UI

> Toolpad works as a drag-and-drop UI builder and uses MUI's own Material UI components. Material UI has almost every component required for internal use cases. Currently, Toolpad supports 16 of them and work on supporting more is continuously ongoing.

Reviewer: ```suggestion
Toolpad works as a drag-and-drop UI builder and uses MUI's own Material UI components. Material UI includes all of the foundational components required for internal use cases. Toolpad currently supports 16 of them, and we're adding more every day.
```

> For additional use-cases, Toolpad supports adding any number of custom components. You can create your own by writing React code. The drag-and-drop builder helps you assemble a quick UI and gives you a head start for the next steps. It ensures **speed and customization**.

Reviewer: ```suggestion
For other use cases, Toolpad supports adding any number of custom React components. The drag-and-drop builder enables you to quickly assemble a minimum viable UI that's endlessly customizable, ensuring the best of both worlds: speed _and_ flexibility.
```

### Connecting to data

Toolpad supports connecting to data in two ways:

> - REST APIs can be easily linked to Toolpad through _HTTP queries_. This is the most common way of getting data to Toolpad.

Reviewer: numbered list, active voice ("do this" instead of "this can be done")

```suggestion
1. Link REST APIs to Toolpad through HTTP queries—this is the most common way of getting data to Toolpad.
```

> - _Custom functions_ where you can write code in your own IDE to connect to any data source. You can reuse your existing scripts, business logic, database models, client libraries, and secrets. These allow you to have a single, shared codebase for internal or external applications meaning **no duplication or fragmentation**.

Reviewer: ```suggestion
2. Write custom functions in your own IDE to connect to any data source. You can reuse your existing scripts, business logic, database models, client libraries, and secrets. These allow you to have a single, shared codebase for internal or external applications—meaning **no duplication or fragmentation**.
```

### Development Lifecycle

> Toolpad ensures you have **full control** over your development lifecycle. Toolpad supports a local-first way of application development. Your project lives on your own file system. You can use your own IDE, version control system, deployment target, and CLI tools.

Reviewer: ```suggestion
Toolpad gives you **full control** over your app's development lifecycle. Toolpad is "local-first": your project lives on your own file system, so you can use your own IDE, version control system, deployment target, and CLI tools.
```

> Like any Node.js app, you can self-host a Toolpad app to your own server or any hosting provider of your choice (AWS, Render, Railway, Heroku etc.). Toolpad does not lock you in to its own cloud hosting.

Reviewer: ```suggestion
Like any Node.js app, you can self-host a Toolpad app on your own server or any hosting provider of your choice (such as AWS, Render, Railway, or Heroku). Toolpad does not lock you in to its own cloud hosting.
```

### Collaboration and Sharing

> The Toolpad application configuration is stored in yaml files that can be checked into git or any version control tool of your choice and maintained collaboratively. We didn't want to reinvent the wheel when it comes to collaboration for engineers. Having a private code repository is one of the essential practices to ensure **safety** of user data.

Reviewer: ```suggestion
The Toolpad application configuration is stored in yaml files that can be checked into git or any version control tool of your choice and maintained collaboratively. We didn't want to reinvent the wheel when it comes to collaboration for engineers. Having a private code repository is one of the essential practices to ensure that user data remains secure.
```

## Conclusion

> Toolpad is a platform to build internal tools that are used for tasks like running marketing campaigns, data analysis, monitoring fraud, updating user data, and more. It aims to reduce the overhead associated with frontend development by providing visual editors and reducing the amount of boilerplate required to bring data on the page. With a drag-and-drop UI, you can build an internal tool for your business logic in just one sitting.

Reviewer: I would move this "Conclusion" text to the "Introduction" 👍

# Why Toolpad?

<p class="description">Learn how Toolpad approaches building internal tools, and how it can be useful to you.</p>

## TL;DR

- **Higher-level API**. Toolpad operates at a higher abstraction level than Material UI. It sets healthy UI constraints that prevent common pitfalls and anti-patterns
- **Drag-and-drop builder**. Toolpad comes with a drag-and-drop builder to simplify UI building. The Drag-and-drop state is persisted in a yaml file so you can fully control the visual output with code
- **Less boilerplate** so you can focus on the important parts of the app
- **Run alongside existing code**. It integrates well with your existing code. You can use your database models, client libraries, secrets, and bespoke components directly
- **Single source of truth**. Because Toolpad runs alongside your existing code, it lets you maintain a single source of truth for all business logic
- **Integrates well with your development lifecycle**. Your project lives on your own file system. You can use your own IDE, version control system, deployment target, CLI tools

## The problems

> After reviewing multiple approaches and tools, we noticed that there are a few recurring core problem areas that needed to be addressed:

Reviewer 2: Based on our style guide, we should avoid using "we": "It’s better to use the passive voice when it allows you to not use “we”"

# Why Toolpad?

<p class="description">Learn how Toolpad approaches building internal tools, and how it can be useful to you.</p>

## TL;DR

- **Higher-level API**. Toolpad operates at a higher abstraction level than Material UI. It sets healthy UI constraints that prevent common pitfalls and anti-patterns.

> - **Drag-and-drop builder**. Toolpad comes with a drag-and-drop builder to simplify UI building. The Drag-and-drop state is persisted in a yaml file so you can fully control the visual output with code.

Reviewer 3: ```suggestion
- **Drag-and-drop builder**. Toolpad comes with a drag-and-drop builder to simplify UI building. The builder saves the UI as a YAML file, so you can even modify and version the visual output on your file system.
```

- **Less boilerplate** so you can focus on the essential parts of the app.
- **Run alongside existing code**. It integrates well with your existing code. You can use your database models, client libraries, secrets, and bespoke components directly.

> - **Single source of truth**. Because Toolpad runs alongside your existing code, it lets you maintain a single source of truth for all business logic.

Reviewer 3: ```suggestion
- **Single source of truth**. Since Toolpad runs alongside your existing code, it lets you maintain a single source of truth for all business logic.
```

- **Integrates well with your development lifecycle**. Your project lives on your file system. You can use your own IDE, version control system, deployment target, CLI tools.

On reviewing multiple products and approaches, a few recurring areas of concern were noticed. Let's examine them in detail:

## Trade-offs with pro-code tools

### Compromised delivery speed

> Every organization aspires to achieve rapid delivery, but to accommodate potential use cases, developers are often compelled to opt for flexible pro-code solutions, which typically come at the expense of speed. These tools offer lower abstraction levels for more control, but handling and scaling them soon becomes an unavoidable maintenance overhead.

Reviewer 3: ```suggestion
Every organization aims to achieve fast delivery, but in order to accommodate various use cases, developers are often pushed to choose flexible pro-code solutions. However, this flexibility often comes at the cost of speed. Pro-code provides lower-level abstractions for greater control, but as use cases scale, they quickly involve a growing maintenance burden that can't be avoided.
```

> ### More lines of code

Reviewer 3: ```suggestion
### A lot of boilerplate
```

> Lots of boilerplate HTML, CSS, and JavaScript glue code is needed just to get backend data to the browser and hooked up to data displaying components.

Reviewer 3: ```suggestion
To connect backend data to the browser and set up data-display components, a substantial amount of boilerplate HTML, CSS, and JavaScript glue code is often required.
```

### Lower-level abstraction

> Building internal tools requires knowledge of front-end framework. While it may interest some, most internal tool developers find the lower-level APIs, HTML and CSS an extra burden to maintain.

Reviewer 3: ```suggestion
Creating internal tools demands knowledge of a front-end framework. While this aspect might be intriguing to some, most developers of internal tools often consider managing the lower-level APIs, HTML, and CSS as an additional burden.
```

## Trade-offs with low-code tools

> There exist many low-code code tools specifically for internal tools. They are proven to be productive but come with some trade-offs, as mentioned below:

Reviewer 3: ```suggestion
There exist many low-code code tools specifically for internal use cases. They offer speed, but come with some trade-offs:
```

> ### Codebase fragmentation and duplication

Reviewer 3: ```suggestion

```

> Frequently, developers invest valuable time in hunting down particular versions of business logic and then maintaining them in multiple places. When working with external low-code/no-code tools, code/logic duplication can often arise as a consequence of design decisions. Not having a unified reference point within a codebase compromises its scalability as well as ease of maintenance and debugging.

Reviewer 3: ```suggestion

```

### Extensibility

> Low-code tools are hard to extend. If there's something the tool can't do out of the box, you can't bring your own code with you and integrate it seamlessly.

Reviewer 3: ```suggestion
Low-code tools can be hard to extend. When the tool can't handle a particular task by default, integrating your own code seamlessly is often a challenge. This lack of flexibility often results in duplicating existing logic or storing your business logic within the configuration files of these apps when trying to add your own code.
```

> 

Reviewer 3: ### Duplication and fragmentation

> 

Reviewer 3: ```suggestion
### Duplication and fragmentation

When working with external low-code or no-code tools, the duplication of code and logic can frequently occur because of a lack of extensibility. Developers invest valuable time in hunting down specific versions of business logic and maintaining them in multiple places. This lack of a unified reference point within the codebase can compromise its scalability, making maintenance and debugging more challenging.
```

### Loss of control

> When working with external tools, developers often forfeit their autonomy and are constrained by the choices offered by the tool. A black-box element intervention takes away the control over the entire process, encompassing integrations, deployment, and version control. This lack of control puts the underlying sensitive data to risk. The security of customer data is critical, and the seriousness of this matter can't be overstated.

Reviewer 3: ```suggestion
When developers use external tools, they frequently sacrifice their autonomy and become restricted by the options provided by the tool. The introduction of a black-box element removes control over the entire process, including integrations, deployment, and version control. This can lead to a significant problem when it comes to the long-term maintainability of internal tools, as it becomes challenging to adapt, extend, or customize them to meet evolving needs.
```

## How Toolpad fits in

Addressing the problems mentioned above leads to a better developer experience. Toolpad solves them in the following ways:

{{"component": "modules/components/DocsImage.tsx", "src": "/static/toolpad/docs/getting-started/why-toolpad/why-toolpad.png", "alt": "How Toolpad fits in your codebase", "caption": "How Toolpad fits in your workflow", "zoom": true }}

### Drag-and-drop builder

Toolpad works as a drag-and-drop UI builder and uses MUI's own Material UI components. Material UI includes all of the foundational components required for internal use cases. Toolpad currently supports 16 of them, and we're continuously adding more.

Toolpad supports adding any number of custom React components for other use cases. The drag-and-drop builder enables you to quickly assemble a minimum viable UI that's endlessly customizable, ensuring the best of both worlds: speed _and_ flexibility.

### Single source of truth

> Toolpad supports connecting to data in two ways:

Reviewer 3: ```suggestion
Toolpad offers two ways to connect to your data, ensuring that you can keep your data in one centralised location and avoid duplicating it for each internal tool you create:
```

> 1. Link REST APIs to Toolpad through HTTP queries—this is the most common way of getting data to Toolpad.

Reviewer 3: ```suggestion
Link REST APIs to Toolpad through HTTP queries: This common method allows you to access your data directly by building REST queries in a visual interface.
```

> 2. Write custom functions in your own IDE to connect to any data source. You can reuse your existing scripts, business logic, database models, client libraries, and secrets. These allow you to have a single, shared codebase for internal or external applications—meaning no duplication or fragmentation.

Reviewer 3: ```suggestion
2. Write custom functions in your own IDE to connect to any data source: You can use your existing scripts, business logic, database models, client libraries, and secrets. This approach helps you maintain a single, shared data source for all your internal tools, preventing data duplication and ensuring consistency across applications.
```

### Integrates with your development lifecycle

> Toolpad gives you complete control over your app's development lifecycle. Toolpad is "local-first"; your project lives on your own file system, so you can use your own IDE, version control system, deployment target, and CLI tools.

Reviewer 3: ```suggestion

Toolpad empowers you with total control over the entire development lifecycle of your application. It operates on a "local-first" principle, where your project resides on your local file system. This enables you to utilize your preferred Integrated Development Environment (IDE), version control system, deployment target, and Command Line Interface (CLI) tools.

Similar to any Node.js application, you have the flexibility to self-host a Toolpad app on your own server or select a hosting provider that suits your needs, whether it's AWS, Render, Railway, or Heroku. Toolpad does not confine you to its cloud hosting, ensuring you have the freedom to choose the hosting solution that best aligns with your preferences and requirements.
```

Like any Node.js app, you can self-host a Toolpad app on your own server or any hosting provider of your choice (such as AWS, Render, Railway, or Heroku). Toolpad does not lock you into its cloud hosting.

### Seamless collaboration

> The Toolpad application configuration is stored in yaml files that can be checked into git or any version control tool of your choice and maintained collaboratively. Having a private code repository is one of the essential practices to ensure that user data remains secure. There is no need to reinvent the wheel regarding collaboration for engineers.

Reviewer 3: ```suggestion
The Toolpad application configuration is saved in YAML files, which can be conveniently added to Git or any preferred version control tool for collaborative maintenance. Maintaining a private code repository is a fundamental practice for enhancing the security of user data. This eliminates the need to create new collaboration mechanisms for engineers, as the existing tools can be effectively leveraged.
```

### Trust and Safety

> Internal tools have access to highly sensitive data that every organization solemnly pledges to safeguard. Despite the stringent security measures in place, data breaches remain a prevalent occurrence. Organizations rigorously engage in due diligence and compliance assessments before trusting a vendor. Toolpad supports just that by letting you keep the control.

Reviewer 3: ```suggestion
While organizations make every effort to protect highly sensitive data and conduct thorough vendor due diligence and compliance assessments, data breaches continue to be a significant concern. Toolpad ensures that you can maintain control over your data, helping to address this issue.
```

> ## How is Toolpad different from Y?

Reviewer 3: Can we remove this section while it is in progress?

# Why Toolpad?

<p class="description">Explains our motivation for building Toolpad, where it differs from other tools and how it could be useful to you.</p>

## Introduction

Toolpad is primarily designed as an application framework for internal tool development. Such applications are often built by teams with access to fewer frontend development resources. These teams face challenges such as:

- Having to keep up with the fast evolving landscape of libraries, frameworks and best practices.
- Lots of glue code is required to get backend data to the browser and hooked up to data displaying components.
- As backends evolve, the internal applications need to evolve with them with minimal amount of maintenance required.

> - Lesser focus on quality assurance, usually lead to higher chances of failure.

Reviewer 3: ```suggestion
- Lesser focus on quality assurance usually leads to higher chances of failure.
```

> ## TLDR

Reviewer 3: ```suggestion
## TL;DR
```

- Integrates well with your existing code. You can reuse your database models, client libraries, secrets, bespoke components
- Integrates well with your development lifecycle. Your project lives on your own file system. You can use your own IDE, version control system, deployment target, CLI tools

> - Comes with a Drag and drop builder to simplify UI building

Reviewer 3: ```suggestion
- Comes with a drag and drop builder to simplify UI building
```

> - Sets Healthy UI constraints that prevent common pitfalls and anti-patterns

Reviewer 3: ```suggestion
- Sets healthy UI constraints that prevent common pitfalls and anti-patterns
```

### Trade-off between speed and customization

Typically, depending on factors such as project size, time constraints, available resources, and scope, developers are compelled to strike a balance between speed and customization. Every business wants features to be delivered at speed, but being able to customize later keeps the project open to come what may. While both aspects are significant, we believe that an architecturally well-designed tool empowers you to attain both objectives simultaneously.

### Trust and Safety

Internal tools have access to highly sensitive data, data that every organization solemnly pledges to safeguard. Despite the stringent security measures in place, data breaches remain a prevalent occurrence. Organizations rigorously engage in due diligence and compliance assessments before placing trust in a vendor. The security of the underlying data remains an exceptionally vital aspect that cannot be underestimated.

## Toolpad

> We think addressing above mentioned problems can lead to a better developer experience. Let' see how Toolpad solves these:

Reviewer 3: ```suggestion
We think addressing the above mentioned problems can lead to a better developer experience. Let's see how Toolpad solves these:
```

### Front end UI

Toolpad works as a drag-and-drop UI builder and uses MUI's own Material UI components. Material UI has almost every component required for internal use cases; currently, Toolpad supports 16 of them and work on supporting more is continuously ongoing.

> For more components, Toolpad supports adding any number of custom components. You can create your own by writing React code. Drag-and-drop builder helps you assemble a quick UI and gives you a head start for the next steps. It ensures **speed and customization**.

Reviewer 3: ```suggestion
For additional use-cases, Toolpad supports adding any number of custom components. You can create your own by writing React code. The drag-and-drop builder helps you assemble a quick UI and gives you a head start for the next steps. It ensures **speed and customization**.
```

### Connecting to data

Toolpad supports connecting to data in two ways:

- REST APIs can be easily linked to Toolpad through _HTTP queries_. This is the most common way of getting data to Toolpad.

> - _Custom functions_ where you can write code in your own IDE to connect to any data source. You can reuse your existing scripts, business logic, database models, client libraries, and secrets files. These allow you to have a single, shared codebase for internal or external applications meaning **no duplication or fragmentation**.

Reviewer 4: ```suggestion
- _Custom functions_ where you can write code in your own IDE to connect to any data source. You can reuse your existing scripts, business logic, database models, client libraries, and secrets. These allow you to have a single, shared codebase for internal or external applications meaning **no duplication or fragmentation**.
```

### Development Lifecycle

Toolpad ensures you have **full control** over your development lifecycle. Toolpad supports a local-first way of application development. Your project lives on your own file system. You can use your own IDE, version control system, deployment target, and CLI tools.

> You can build your app and deploy it like any other Node.js app, to any hosting provider of your choice (AWS, Render, Railway, Heroku etc.). Toolpad does not lock you in to its own cloud hosting.

Reviewer 4: We should probably also mention self-hosting, on your own servers?

### Collaboration and Sharing

> A Toolpad app is a set of files that can be simply added to a GitHub/GitLab repo and maintained collaboratively. We didn't want to reinvent the wheel when it comes to collaboration for engineers. Having a private code repository is one of the essential practices to ensure **safety** of user data.

Reviewer 4: ```suggestion
The Toolpad applicationconfiguration is stored in yaml files that can be checked into your version control tools such as git and maintained collaboratively. We didn't want to reinvent the wheel when it comes to collaboration for engineers. Having a private code repository is one of the essential practices to ensure **safety** of user data.
```

<!---
### Consistent Design

Maintaining a consistent design language throughout a web application is crucial for a seamless user experience. Toolpad helps you achieve this by providing pre-built components that follow Material Design guidelines.

Toolpad also supports adding a theme to a page which can help with a UI that matches your brand guidelines. The components have individual overrides as well, in case you wish to customize them further.

### Open-Source

MUI Toolpad is an open-source project, which follows an open-core model. It means that the core is going to remain free to use forever. In the future, we may add add paid features to support advanced use cases. However, we will not attempt to monetise the project via artificial constraints on usage, such as limits on the number of connections, components or queries.

The open-source nature of Toolpad also means that it is constantly being updated and improved by the MUI team and the community.
-->

## Conclusion

> Toolpad is a platform to build internal tools that are used for tasks like running marketing campaigns, data analysis, monitoring fraud, updating user data, and more. It aims to reduce the overhead associated with frontend development by providing visual editors and reducing the amount of boilerplate required to bring data on the page. With basic frontend knowledge you can build an app in just one sitting.

Reviewer 3: ```suggestion
Toolpad is a platform to build internal tools that are used for tasks like running marketing campaigns, data analysis, monitoring fraud, updating user data, and more. It aims to reduce the overhead associated with frontend development by providing visual editors and reducing the amount of boilerplate required to bring data on the page. With a drag-and-drop UI, you can build an internal tool for your business logic in just one sitting.
```

# Why Toolpad?

<p class="description">Explains our motivation for building Toolpad, where it differs from other tools and how it could be useful to you.</p>

## Introduction

> Toolpad is a developer tool primarily for the admin applications use case. Building an admin panel (aka internal tools) can be intimidating because:

Reviewer 4: Maybe something more along the lines of

```suggestion
Toolpad is primarily designed as an application framework for internal tool development. Such application are often built by teams with access to fewer frontend development resources. These teams face challenges such as:

- Having to keep up with the fast evolving landscape of libraries, frameworks and best practices.
- Lots of glue code is required to get backend data to the browser and hooked up to data displaying components.
- As backends evolve, the internal applications need to evolve with them with minimal amount of maintenance required.
- ...?
```

?

- it is not focused on frontend
- it is difficult to keep up with the fast evolving landscape of frameworks, libraries and best practices
- lots of glue code is required to display the data
- it requires connecting to multiple data sources

> - security of the underlying sensitive data needs to ensured at all times

Reviewer 4: This isn't exactly specific to internal applications, personally I'd leave it out

```suggestion
```

MUI Toolpad aims to reduce the overhead associated with frontend development by providing visual editors and reducing the amount of boilerplate required to bring data on the page. With basic frontend knowledge you can build an app in just one sitting. On this page, we'll try to address your concerns by covering all aspects of building an admin panel. This shall help you evaluate if Toolpad is the right product for you.

## TLDR

- **Integrates well with your existing code**. You can reuse your database models, client libraries, secrets, bespoke components
- Integrates well with your **development lifecycle.** Your project lives on your own file system. You can use your own IDE, version control system, deployment target, CLI tools
- Comes with a **Drag and drop builder** to simplify UI building
- Sets **Healthy UI constraints** that prevent common pitfalls and anti-patterns
- Lets you maintain a **single source of truth** for all business logic
- **Less boilerplate** so you can focus on the important parts of the app

## Admin builders market overview

> The main problem with the tools currently in the Node.js and TypeScript ecosystem is that they require a major tradeoff between flexibility and speed. For a back office application developer, the backend and the database schema is an area of comfort. What's relatively challenging is building the UI and having a reasonable UX that can show the required data in a readable and editable format. And that's where all the admin builders come into the picture. If we closely look at the market, we can divide them into the following categories:

Reviewer 3: ```suggestion
The main problem with the tools currently in the  ecosystem is that they require a major tradeoff between flexibility and speed. For a backend developer working on internal tools, the bottleneck usually is rapidly creating minimally usable UIs for these tasks. That is the primary problem being solved by admin builders. If we look closely at the market, we can divide them into the following categories:
```

### Fast development, less flexibility

> Most modern internal tool builders in the market fall into this category. Products like Retool, Appsmith act like a SaaS and provide a UI for rapid application development. The apps can be either hosted by you or them. They are excellent products for someone who's looking to build fast. The caveats are that you can't use your IDE, your existing codebase, or logic, and it can cost a lot as you scale.

Reviewer 3: ```suggestion
Most modern internal tool builders in the market fall into this category. Products like Retool, Appsmith act like a SaaS and provide a UI for rapid application development. The apps can either be hosted by you or them. They are excellent products for someone who's looking to build fast. The caveats are that you can't use your IDE, your existing codebase or custom scripts, and it can cost a lot as you scale.
```

### Moderate development speed, moderate flexibility

> Products like airplane.dev, interval.com fall under this category. They have APIs for all admin use cases; while this can be great for those who enjoy writing code, it is a bit too much as adding a component also requires writing code. Their main advantage over previous category is that they store their configuration in local code files. This puts more control over the development lifecycle back in the developer's hands. Users can version control, deploy, and use any IDE they want.

Reviewer 3: ```suggestion
Products like airplane.dev, interval.com fall under this category. They require users to write code to describe their UIs. The main advantage in this category over previous ones is that they store configuration in local code files.  This allows users more control over the development lifecycle, including things like version control and integration with your codebase.
```

### Slow development speed, more flexibility

> Products like react admin, refine.dev fall in this category, they are all code and have APIs for everything to build an admin panel from scratch. They don't support cloud hosting and are a relatively cheaper alternative.

Reviewer 3: ```suggestion
Products like react-admin, refine.dev fall in this category. They involve writing code to build an admin panel from scratch. There is no UI editor and  cloud hosting.
```

> We'll do a detailed comparison with these tools in the future.

Reviewer 3: Why mention this? We can either do the comparison now and add it, or just add the comparison later?

## Toolpad

> How is Toolpad different from the above tools, and why should you use it? Toolpad is a **flexible and fast** way to build a back office application. Let's see how:

Reviewer 3: ```suggestion
How is Toolpad different from the above tools, and why should you use it? Toolpad is a **flexible and fast** way to build a internal tool. Let's see how:
```

### Front end UI

> Toolpad works as a drag-and-drop UI builder and uses MUI's own Material UI components. Material UI has almost every component required for back-office use cases; currently, 16 of them are there in Toolpad but we are always adding more.

Reviewer 3: ```suggestion
Toolpad works as a drag-and-drop UI builder and uses MUI's own Material UI components. Material UI has almost every component required for internal use cases; currently, Toolpad supports 16 of them and work on supporting more is continuously ongoing.
```

For more components, Toolpad supports adding any number of custom components. You can create your own by writing React code.

Drag-and-drop builder helps you assemble a quick UI and gives you a head start for the next steps.

### Consistent Design

> Maintaining a consistent design language throughout a web application is crucial for a seamless user experience. Toolpad helps you achieve this by providing pre-built components that follow Material Design guidelines. This ensures your application looks consistent across all pages and screens.

Reviewer 3: ```suggestion
Maintaining a consistent design language throughout a web application is crucial for a seamless user experience. Toolpad helps you achieve this by providing pre-built components that follow Material Design guidelines.
```

> Toolpad also supports adding a theme to a page, it can help with a UI that matches your brand guidelines. The components have overrides in case wish to customize them further.

Reviewer 3: ```suggestion
Toolpad also supports adding a theme to a page which can help with a UI that matches your brand guidelines. The components have individual overrides as well, in case you wish to customize them further.
```

### Connecting to data

> Imagine you've built the front end, the next step is connecting your data. In Toolpad, you can bring your entire data from your existing backend in two ways:

Reviewer 3: ```suggestion
Once you've built the front end, the next step is connecting to your data. In Toolpad, you can bring data from your existing backend in two ways:
```

> - REST APIs can be easily linked to Toolpad through **HTTP queries**. This is the most common way of getting data to Toolpad

Reviewer 3: ```suggestion
- REST APIs can be easily linked to Toolpad through **HTTP queries**. This is the most common way of getting data to Toolpad.
```

> - **Custom functions** where you can write code in your own IDE to connect to any database. If there is business logic in your customer-facing application that could be useful in Toolpad; you can simply import it into Toolpad. You can reuse your database models, client libraries, and secrets. However, remember that currently only Node.js backend is supported in Toolpad.

Reviewer 3: ```suggestion
- **Custom functions** where you can write code in your own IDE to connect to any data source. You can reuse your existing scripts, business logic, database models, client libraries, and secrets files. However, remember that currently, Toolpad only supports Node.js functions
```

> Later we'll also support REST API like UI for graphQL, Open API, etc.

Reviewer 3: I suggest not making commitments in docs pages that we might not remember to maintain. Instead, we can update this with the link to the public roadmap once that is ready.

### Development Lifecycle

Toolpad ensures you have full control over your development lifecycle. Toolpad supports a local-first way of application development. Your project lives on your own file system. You can use your own IDE, version control system, deployment target, and CLI tools.

### Deployment

> After building an app, you might wonder where to put it. Toolpad allows you to host your app on any platform of your choice. Currently, we don't offer cloud hosting. Instead, we suggest self-hosting like how you'd have done for your customer-facing application.

Reviewer 3: ```suggestion
You can build your app and deploy it to any hosting provider of your choice. Toolpad does not lock you in to its own cloud hosting. 
```

> A Toolpad app is hosted like any other Node.js project. You can choose from providers like AWS, Render, Heroku, etc., without hassle.

Reviewer 3: ```suggestion

```

### Collaboration and Sharing

> MUI Toolpad is an ideal tool for teams as it allows collaboration with others through a shared codebase, through established practices like Git. We didn't want to reinvent the wheel and wanted to support already accepted ways of code collaboration. A Toolpad app is a set of files that can be simply put on a GitHub/GitLab repo and maintained collaboratively.

Reviewer 3: ```suggestion
A Toolpad app is a set of files that can be simply added to a GitHub/GitLab repo and maintained collaboratively. We didn't want to reinvent the wheel when it comes to collaboration for engineers.
```

### Open-Source

> MUI Toolpad is an open-source project, which follows an open-core model. It means that the core is going to remain free to use forever. We'll add paid features to support advanced use cases in the future. We won't put any artificial limits like the number of apps, data connections, components, etc.

Reviewer 3: ```suggestion
MUI Toolpad is an open-source project, which follows an open-core model. It means that the core is going to remain free to use forever. In the future, we may add add paid features to support advanced use cases. However, we will not attempt to monetise the project via artificial constraints on usage, such as limits on the number of connections, components or queries.
```

> The open-source nature of Toolpad also means that it is constantly being updated and improved by the MUI team and the community. This ensures that it remains up-to-date with the latest trends and technologies.

Reviewer 3: ```suggestion
The open-source nature of Toolpad also means that it is constantly being updated and improved by the MUI team and the community.
```

## Conclusion

> Toolpad is a platform to build behind-the-scenes admin tools that are used for tasks like running marketing campaigns, checking daily data and stats, reporting fraud, updating user records, and more. These back office tools are more about function and less about design.

Reviewer 3: ```suggestion
Toolpad is a platform to build internal tools that are used for tasks like running marketing campaigns, data analysis, monitoring fraud, updating user data, and more. These tools often prioritise function over form.
```

> When building such an app, developers have certain expectations from the tools they use. Some prefer the old way of starting from scratch to maintain control over every detail - like if a new button needs to be added in a particular spot with a specific look. If they rely on tools like Toolpad, there might be differences in execution but rest assured, it can handle most custom scenarios efficiently.

Reviewer 3: ```suggestion

```

> ### The problems with pro-code tools

Reviewer 4: I'm not sure whether we need to talk about "problems with pro-code" as we talk about "problems with low-code". The way I see it is not that there are problems with pro-code. It's that for internal tooling there are different trade-offs you can make. e.g. trade in flexibility, and performance for speed of development and maintenance overhead.
